src/sentry/dynamic_sampling/latest_release_booster.py

```diff
@@ -1,234 +1,382 @@
+            # release with a different environment.
+            #
+            # In theory we could add a check that verifies whether the incoming release is already being boosted and
+            # only has a different environment and in this case use only the > but it adds complexity for no big benefit.
+            if latest_release_date is None or incoming_release_date >= latest_release_date:
+                self._update_latest_release_date(timestamp=incoming_release_date)
+                return True
 
-    if expired_releases:
-        redis_client.hdel(cache_key, *expired_releases)
+        return False
 
-    return boosted_releases
+    def _update_latest_release_date(self, timestamp: float) -> None:
+        cache_key = self._generate_cache_key_for_project_latest_release()
+        self.redis_client.set(cache_key, timestamp)
 
+    def _get_release_date_from_incoming_release(self) -> Optional[float]:
+        release = self.latest_release_params.release
 
-def add_boosted_release(project_id: int, release_id: int, environment: Optional[str]) -> None:
-    """
-    Function that adds a release to the list of active boosted releases for a given project.
-    """
-    # Called here for expired releases cleanup
-    get_boosted_releases(project_id)
-
-    cache_key = generate_cache_key_for_boosted_releases_hash(project_id)
-    redis_client = get_redis_client_for_ds()
-    # TODO(ahmed): Modify these two statements into one once we upgrade to a higher redis-py version as in newer
-    #  versions these two operations can be done in a single call.
-    redis_client.hset(
-        cache_key,
-        generate_cache_key_for_boosted_release_with_environment(release_id, environment),
-        datetime.utcnow().replace(tzinfo=UTC).timestamp(),
-    )
-    redis_client.pexpire(cache_key, BOOSTED_RELEASE_TIMEOUT * 1000)
-
-
-def get_augmented_boosted_releases(project_id: int, limit: int) -> List[ExtendedBoostedRelease]:
+        if release.date_released:
+            return float(release.date_released.timestamp())
+        elif release.date_added:
+            return float(release.date_added.timestamp())
+
+        return None
+
+    def _get_release_date_from_latest_release(self) -> Optional[float]:
+        cache_key = self._generate_cache_key_for_project_latest_release()
+        timestamp = self.redis_client.get(name=cache_key)
+        return float(timestamp) if timestamp else None
+
+    def _generate_cache_key_for_project_latest_release(self) -> str:
+        return f"ds::p:{self.latest_release_params.project.id}:latest_release"
+
+    def _generate_cache_key_for_observed_release(self) -> str:
+        return (
+            f"ds::p:{self.latest_release_params.project.id}"
+            f":r:{self.latest_release_params.release.id}"
+            f"{_get_environment_cache_key(self.latest_release_params.environment)}"
+        )
+
+
+class LatestReleaseBooster:
```

Comment:
Why is this a class? As far as I can tell, it only lives as a temporary when we call `observe_release().boost_if_not_observed()`, so why not inline the logic of this class in `observe_release()`?

Reply:
I can inline the logic, for sure, it was more a stylistic decision to decouple the two operations. If you feel like it might improve the code I will do the refactor.

Reply:
yeah agree with @jjbayer